examples/lib/stories/collision_detection/raytrace_example.dart

```diff
@@ -0,0 +1,184 @@
+      CircleComponent(
+        radius: 500,
+        paint: boxPaint,
+        children: [CircleHitbox()],
+      ),
+    ]);
+  }
+
+  final extraChildren = <Component>[];
+  @override
+  void onTap() {
+    if (extraChildren.isEmpty) {
+      addAll(
+        extraChildren
+          ..addAll(
+            [
+              CircleComponent(
+                position: Vector2(100, 100),
+                radius: 50,
+                paint: boxPaint,
+                children: [CircleHitbox()],
+              ),
+              CircleComponent(
+                position: Vector2(150, 500),
+                radius: 50,
+                paint: boxPaint,
+                anchor: Anchor.center,
+                children: [CircleHitbox()],
+              ),
+              CircleComponent(
+                position: Vector2(150, 500),
+                radius: 150,
+                paint: boxPaint,
+                anchor: Anchor.center,
+                children: [CircleHitbox()],
+              ),
+              RectangleComponent(
+                position: Vector2.all(300),
+                size: Vector2.all(100),
+                paint: boxPaint,
+                children: [RectangleHitbox()],
+              ),
+              RectangleComponent(
+                position: Vector2.all(500),
+                size: Vector2(100, 200),
+                paint: boxPaint,
+                children: [RectangleHitbox()],
+              ),
+              CircleComponent(
+                position: Vector2(650, 275),
+                radius: 50,
+                paint: boxPaint,
+                anchor: Anchor.center,
+                children: [CircleHitbox()],
+              ),
+              RectangleComponent(
+                position: Vector2(550, 200),
+                size: Vector2(200, 150),
+                paint: boxPaint,
+                children: [RectangleHitbox()],
+              ),
+              RectangleComponent(
+                position: Vector2(350, 30),
+                size: Vector2(200, 150),
+                paint: boxPaint,
+                angle: tau / 10,
+                children: [RectangleHitbox()],
+              ),
+            ],
+          ),
+      );
+    } else {
+      removeAll(extraChildren);
+      extraChildren.clear();
+    }
+  }
+
+  @override
+  void onMouseMove(PointerHoverInfo info) {
+    final origin = info.eventPosition.game;
+    isOriginCasted = origin == this.origin;
+    this.origin = origin;
+  }
+
+  final Ray2 _ray = Ray2.zero();
+  var _timePassed = 0.0;
+
+  @override
+  void update(double dt) {
+    super.update(dt);
+    _timePassed += dt;
+    rayPaint.color = _colorTween.transform(0.5 + (sin(_timePassed) / 2))!;
+    if (origin != null && !isOriginCasted) {
+      _ray.origin.setFrom(origin!);
+      _ray.direction
+        ..setValues(-1, 1)
+        ..normalize();
+      const angle = tau / numberOfRays;
+      for (var i = 0; i < numberOfRays; i++) {
+        _ray.direction.rotate(i * angle);
```

Comment:
I had it like that first, the problem is that the ray's direction can still be modified in place and then you'd have unexpected results. I had it like a `NotifyingVector2` too, but that was too heavy.

Reply:
But you'd also have unexpected results if you modify in-place and _do not_ call `updateNormals()`. So, if the user calls `updateNormals`, they already know they are doing something iffy here, and so they might as well use the regular setter instead.

This kinda is a global problem for Flame. Whenever we have an object with some property that is a `Vector2`, then exposing that property to the user may result in the user modifying that vector in-place, which could break some important internal constraints. I can see several solutions here:
1. Create `UnmodifiableVector2` which would be a read-only view on the underlying vector, and return this object. 
2. Return a clone of the underlying vector (perhaps clone into a static "temporary" `Vector2`).
3. Return the underlying vector as-is, but have an implicit agreement with the user that they would not modify that vector. This option is most efficient but least safe.
4. Return a `NotifyingVector2` and register a callback that would update the internal state in case the user changes the vector. This approach is currently used in `PositionComponent`, and it seems to be the most expensive one.

Reply:
I think only 1 and 3 are possible here (like I said, I already tried 4). I think I'll go with 3 and then and in a later PR we can possibly introduce 1 on multiple places.

Reply:
Did 3